uefi/src/proto/misc.rs

```diff
@@ -23,3 +29,55 @@ impl Timestamp {
         unsafe { (self.0.get_properties)(&mut properties) }.to_result_with_val(|| properties)
     }
 }
+
+/// Protocol to register for a notification when ResetSystem is called.
+#[derive(Debug)]
+#[repr(transparent)]
+#[unsafe_protocol(ResetNotificationProtocol::GUID)]
+pub struct ResetNotification(ResetNotificationProtocol);
+
+impl ResetNotification {
+    /// Register a notification function to be called when ResetSystem() is called.
+    ///
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use log::info;
+    /// use uefi::Handle;
+    /// use uefi::prelude::BootServices;
+    /// use uefi::proto::misc::{ResetNotification};
+    /// use uefi_raw::Status;
+    /// use uefi_raw::table::runtime;
+    ///
+    ///
+    /// // value efi_reset_fn is the type of ResetSystemFn, a function pointer
+    /// unsafe extern "efiapi" fn efi_reset_fn(
+    ///         rt: runtime::ResetType,
+    ///         status: Status,
+    ///         data_size: usize,
+    ///         data: *const u8,
+    /// ){
+    ///     info!("Inside the event callback");
+    ///     info!("do what you want");
+    /// }
+    ///
+    /// pub fn test(image: Handle, bt: &BootServices) {
+    ///
+    ///     let mut rn = bt
+    ///         .open_protocol_exclusive::<ResetNotification>(image)
+    ///         .expect("Failed to open Timestamp protocol");
+    ///
+    ///     rn.register_reset_notify(Some(efi_reset_fn))
+    ///         .expect("Failed to register a reset notification function!");
+    /// }
+    /// ```
+    pub fn register_reset_notify(&mut self, reset_function: Option<ResetSystemFn>) -> Result {
```

Comment:
According to the [spec](https://uefi.org/specs/UEFI/2.10/39_Micellaneous_Protocols.html#efi-reset-notification-protocol-registerresetnotify), register and unregister will return EFI_INVALID_PARAMETER if the reset function is null, so we just shouldn't allow that in the safe API. Drop the `Option` so that the caller is forced to provide a valid input.